### PR TITLE
How to quantize document

### LINKED ISSUE #931: Documentation: How to quantize
Blueoil's main feature is binary quantize.
For quantize, We need to quantize both Weight and Activation.
It's good to write how to quantize a network.

### PR DESCRIPTION
**Pull request by oatawa1**
## What this patch does to fix the issue.
Create a quantize document
Require #954 to make equation render correctly

## Link to any relevant issues or pull requests.
Close #931 

### Overall
- [X] How to quantize the Weight
- [X] How to quantize the Activation
- [x] How to use TensorFlow's custom getter and variable_scope to quantize
- [x] Write simple sample code in your documentation.
- [x] Insert in to index
### Ops
- [X] BinaryChannelWiseMeanScalingQuantizer
- [X] BinaryMeanScalingQuantizer
- [X] LinearMidTreadHalfQuantizer

### CODE REVIEW

**Comment by oatawa1** [comment]
@iizukak 
Thanks, I moved to the new section and now eq. is rendered correctly.

I changed it to PR now. Could you review this?

File: docs/quantize/quantize_example.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+from functools import partial
+
+import tensorflow as tf
+
+from blueoil.networks.<TASK_NAME>.<NETWORK_NAME> import FooNetwork
```
Comment (iizukak) [inline docs/quantize/quantize_example.py:5]: We want a small example network. Not template.
It's ok to assume task as classification in this case.

File: docs/quantize/quantize_example.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+
+class FooNetworkQuantize(FooNetwork):
```
Comment (iizukak) [inline docs/quantize/quantize_example.py:8]: In example code, `FooNetworkQuantize` should inherit [`Base` ](https://github.com/blue-oil/blueoil/blob/master/blueoil/networks/classification/base.py). Not `FooNetwork`.
We assume the document user want to write their network from scratch.

File: docs/quantize/quantize_example.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    """Quantize Foo Network."""
+
+    def __init__(
+            self,
+            quantize_first_convolution=True,
```
Comment (iizukak) [inline docs/quantize/quantize_example.py:13]: Please reduce these arguments.
In example, it's ok to fix these parameters.

File: docs/quantize/quantize_example.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
+from functools import partial
```
Comment (iizukak) [inline docs/quantize/quantize_example.py:1]: Umm, Sorry, This file may should put under the  `blueoil/networks/` directory.
And it's good to separate to 2 PRs. 

1. Example code PR.
2. Documentation PR.

File: docs/quantize/weight_quantization.rst [hunk 5]
```diff
@@ -0,0 +1,71 @@
+Weight quantization
```
Comment (iizukak) [inline docs/quantize/weight_quantization.rst:1]: Wow, nice documentation.

File: docs/quantize/activation_quantization.rst [hunk 6]
```diff
@@ -0,0 +1,47 @@
+Activation quantization
+=======================
+
+Blueoil quantize an activation function when the last convolution layer is quantized. (``quantize_last_conv`` =True), otherwise ``Leaky ReLU`` is used as an activation function.
```
Comment (iizukak) [inline docs/quantize/activation_quantization.rst:4]: Let me check my understand, If `quantize_last_conv = True` What happen?
`quantize_last_convolution` variable is really used?

File: docs/quantize/quantize_example.py (hunk 2 above)
Comment (oatawa1) [reply docs/quantize/quantize_example.py:8]: I think we need `FooNetwork` which inherit from `Base`.
Then `FooNetwotkQuantize` inherit from `FooNetwork`.

I will add the FooNetwork code in this file.

File: docs/quantize/activation_quantization.rst (hunk 6 above)
Comment (oatawa1) [reply docs/quantize/activation_quantization.rst:4]: Sorry, my typo the variable name should be `quantize_last_convolution`

I think it used in quantized network class constructor.
 ```
def __init__( ..., quantize_last_convolution, ):
        self.quantize_last_convolution = quantize_last_convolution
        ...
        self.activation = activation_quantizer(**activation_quantizer_kwargs)
        ...
        if self.quantize_last_convolution:
                self.before_last_activation = self.activation
        else:
                self.before_last_activation = lambda x: tf.nn.leaky_relu(x, alpha=0.1, name="leaky_relu")
 ```

Ref:
`lm_fyolo.py` https://github.com/blue-oil/blueoil/blob/ac416513e1ba8fccd92742b68e9c8a807f1e54ff/blueoil/networks/object_detection/lm_fyolo.py#L188
`darknet.py` https://github.com/blue-oil/blueoil/blob/ac416513e1ba8fccd92742b68e9c8a807f1e54ff/blueoil/networks/classification/darknet.py#L306

File: docs/quantize/quantize_example.py (hunk 4 above)
Comment (oatawa1) [reply docs/quantize/quantize_example.py:1]: 1. will be in #966

File: docs/quantize/quantize_example.py (hunk 1 above)
Comment (oatawa1) [reply docs/quantize/quantize_example.py]: Please review this in #966 instead.

Comment (oatawa1) [reply docs/quantize/quantize_example.py]: This change to classification task and put in `blueoil/networks/classification` in #966 .

File: docs/quantize/weight_quantization.rst (hunk 5 above)
Comment (oatawa1) [reply docs/quantize/weight_quantization.rst:1]: Thanks.

File: docs/quantize/quantize_example.py (hunk 2 above)
Comment (iizukak) [reply docs/quantize/quantize_example.py:49]: I'm not sure why FooNetwotkQuantize can not directory inherit Base?

File: docs/quantize/quantize_example.py (hunk 1 above)
Comment (iizukak) [reply docs/quantize/quantize_example.py]: Thanks!! I'll review the sample code PR.

**Comment by oatawa1** [comment]
@iizukak Thanks for the review. 

Currently, I'll keep `quantize_example.py` in this PR until #966 merge.

File: docs/quantize/activation_quantization.rst (hunk 6 above)
Comment (oatawa1) [reply docs/quantize/activation_quantization.rst]: Oh I found that, the current config file did not contain `quantize_last_convolution`, I will check this again.

**Comment by iizukak** [comment]
@oatawa1 
I want quantized graph structure.
How about to add netron image of sample network?

**Comment by oatawa1** [comment]
@iizukak Thank you for the suggestion, I added the graph structure image.

**Comment by iizukak** [comment]
@oatawa1 Can you remote WIP now?

**Comment by oatawa1** [comment]
@iizukak I'm sorry I don't understand what you mean?

I'm working on this PR now and test on the server.

**Comment by iizukak** [comment]
@oatawa1 
About PR title. 
WIP means "not ready for merge".
Is this PR ready for merge?

**Comment by oatawa1** [comment]
@iizukak Understood. 
This PR is not ready to merge.

I will notify you again when this PR is ready.

**Comment by iizukak** [comment]
@oatawa1 I see 😄

**Comment by oatawa1** [comment]
@iizukak I removed `[WIP]` from the title. Please review this PR again.

**Comment by oatawa1** [comment]
@iizukak Thanks.

**Pull request merged**